src/Middleware/HttpLogging/src/W3CLoggingMiddleware.cs

```diff
@@ -0,0 +1,194 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Options;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.HttpLogging
+{
+    /// <summary>
+    /// Middleware that logs HTTP requests and HTTP responses.
+    /// </summary>
+    internal sealed class W3CLoggingMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly W3CLogger _w3cLogger;
+        private readonly IOptionsMonitor<W3CLoggerOptions> _options;
+
+        /// <summary>
+        /// Initializes <see cref="W3CLoggingMiddleware" />.
+        /// </summary>
+        /// <param name="next"></param>
+        /// <param name="options"></param>
+        /// <param name="environment"></param>
+        public W3CLoggingMiddleware(RequestDelegate next, IOptionsMonitor<W3CLoggerOptions> options, IHostEnvironment environment)
+        {
+            _next = next ?? throw new ArgumentNullException(nameof(next));
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            if (environment == null)
+            {
+                throw new ArgumentNullException(nameof(environment));
+            }
+
+            _options = options;
+            if (string.IsNullOrEmpty(_options.CurrentValue.LogDirectory))
+            {
+                // Logs are written in the app directory in a folder named 'logs/{UTC Timestamp}'
+                _options.CurrentValue.LogDirectory = Path.Join(environment.ContentRootPath, "logs", DateTimeOffset.Now.ToUnixTimeSeconds().ToString(CultureInfo.InvariantCulture));
+            }
+            _w3cLogger = new W3CLogger(_options);
+        }
+
+        /// <summary>
+        /// Invokes the <see cref="HttpLoggingMiddleware" />.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>HttpResponseLog.cs
+        public async Task Invoke(HttpContext context)
+        {
+            var options = _options.CurrentValue;
+
+            var w3cList = new List<KeyValuePair<string, object?>>();
+
+            if (options.LoggingFields.HasFlag(W3CLoggingFields.Date) || options.LoggingFields.HasFlag(W3CLoggingFields.Time))
+            {
+                AddToList(w3cList, nameof(DateTime), DateTime.Now.ToString(CultureInfo.InvariantCulture));
+            }
+
+            if (options.LoggingFields.HasFlag(W3CLoggingFields.UserName))
+            {
+                AddToList(w3cList, nameof(HttpContext.User), context.User is null ? "" : (context.User.Identity is null ? "" : (context.User.Identity.Name is null ? "" : context.User.Identity.Name)));
```

Comment:
I assume this would be one of the first middleware in the pipeline, if so in that state there's a high chance that `HttpContext.User` is null. Is it possible to move add username after calling `next`?

Reply:
Sure, I'll move it